kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -107,6 +119,70 @@ public Row getScanState(TableClient tableClient) {
 
     @Override
     public Optional<Predicate> getRemainingFilter() {
-        return filter;
+        return getDataFilters();
+    }
+
+    private Optional<Tuple2<Predicate, Predicate>> splitFilters(Optional<Predicate> filter) {
+        return filter.map(predicate -> {
+            Metadata metadata = protocolAndMetadata.get()._2;
+            Set<String> partColNames = metadata.getPartitionColumns().stream().collect(toSet());
+            return PartitionUtils.splitMetadataAndDataPredicates(predicate, partColNames);
+        });
+    }
+
+    private Optional<Predicate> getDataFilters() {
+        return removeAlwaysTrue(partitionAndDataFilters.get().map(filters -> filters._2));
+    }
+
+    private Optional<Predicate> getPartitionsFilters() {
+        return removeAlwaysTrue(partitionAndDataFilters.get().map(filters -> filters._1));
+    }
+
+    /**
+     * Consider `ALWAYS_TRUE` as no predicate.
+     */
+    private Optional<Predicate> removeAlwaysTrue(Optional<Predicate> predicate) {
+        return predicate.map(filter -> {
+            if (filter.getName().equalsIgnoreCase("ALWAYS_TRUE")) {
+                return null;
+            }
+            return filter;
+        });
+    }
+
+    private CloseableIterator<FilteredColumnarBatch> applyPartitionPruning(
+        TableClient tableClient,
+        CloseableIterator<FilteredColumnarBatch> scanFileIter) {
+        Optional<Predicate> partitionPredicate = getPartitionsFilters();
+        if (!partitionPredicate.isPresent()) {
+            // There is no partition filter, return the scan file iterator as is.
+            return scanFileIter;
+        }
+
+        Metadata metadata = protocolAndMetadata.get()._2;
+        Set<String> partitionColNames = metadata.getPartitionColumns().stream().collect(toSet());
+        Map<String, DataType> partitionColNameToTypeMap = metadata.getSchema().fields().stream()
+            .filter(field -> partitionColNames.contains(field.getName()))
+            .collect(Collectors.toMap(
```

Comment:
nit: is this usage inconsistent? `toSet` vs `Collectors.toMap`?

Reply:
Ok, changed to `toMap`.